README.md
# Easy_GA
[![Current Crates.io Version](https://img.shields.io/crates/v/easy_ga.svg)](https://crates.io/crates/easy_ga)

Easy_GA is a genetic algorithm library made for Rust projects. It provides full customization for your own genotypes definitions and a geneticAlgorithm implementation to wrap all the common logic within a genetic algorithm.

All the changes will be updated in the [CHANGELOG.md](CHANGELOG.md)

## Features
- `trait Gene`: Definition to implement for your custom genotypes.
- `trait Selection`: Definition for your custom selection algorithms.
    * `Roulette`: Selection algorithm already implemented.
    * `Tournament`: Selection algorithm implementation with `n` members on it.
    * `Random`: Selection algorithm already implemented.
    * `Stochastic`: Selection algorithm already implemented.
- `GeneticAlgorithm`: The main class to wrap the business logic in the genetic algorithm execution.

## Usage

In your `Cargo.tml` you have to add the `Easy_GA` dependency
```rust
[dependencies]
easy_ga = "*"
```
---
Now I will show you a basic example of `Easy_GA` that you can find on [main.rs](src/main.rs)

Files to include in order to use features:

```rust
use easy_ga::Gene; // For defining our own gene.
use easy_ga::GeneticAlgorithm; // To create a GeneticAlgorithm.
use easy_ga::SelectionAlgorithms; // To specity a concrete SelectionAlgorithm.
```
---
Definition of a custom Gene implementing `easy_ga::Gene` trait:

```rust
#[derive(Clone, Copy)]
struct MyGene {
    // Fields.
    fitness: f64 // Recomended to avoid recalculate fitness on `get_fitness`
}

impl Gene for MyGene {
    fn init() -> Self {
        // Gene constructor.
    }

    fn calculate_fitness(&mut self) -> f64 {
        // Fitness function.
    }

    fn crossover(&self, other: &Self) -> Self {
        // Crossover implementation.
    }

    fn mutate(&mut self) {
        // Mutation implementation.
    }

    fn get_fitness(&self) -> f64 {
        // Returns the fitness
    }
}
```
At this moment, we need to implement the `Clone` & `Copy` traits for our `Gene`. I will try to avoid that in a future versions.

---

Initialization of our `GeneticAlgorithm`:

```rust
let genetic_algorithm = GeneticAlgorithm::<MyGene>::new()
            .population_size(20)
            .iterations(50)
            .mutation_rate(0.10)
            .selection_rate(0.90)
            .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(10)))
            .fitness_goal(100.0)
            .init().unwrap();
```
We have other ways to initializate our `GeneticAlgorithm` such as `GeneticAlgorithm::new_with_values` if we don't want the chain calling method.

---

Now that we have defined our genotype and have initializate our `GeneticAlgorhtm` we have 2 ways of running it:

- __`GeneticAlgorithm::run`__: This method runs the algorithm until the end and returns a tuple with (`Gene`, `StopCriteria`) that represents the best `Gene` in the execution and the reason to stop the execution.
```rust
let (gene, stop_criteria) = genetic_algorithm.run();
```
- __Iteration by iteration__: We have the posibilty of running the algorithm generation by generation and make modification while the execution is running.
```rust
while genetic_algorithm.is_running() {
    let new_generation: &Vec<MyGene> = genetic_algorithm.next_iteration();
}
```
---
## Logger

The logger is a very usefull tool to measure and retrieve some data from the execution. By default the logger is disabled, you can enable it this way:

```rust
use easy_ga::VerbosityLevel; // Verbosity level {DISABLED, LOW, MID, HIGH}
use easy_ga::VerbosityType; // Verbosity type {LOG, SAVE, LOG_AND_SAVE}
use easy_ga::LOG_verbosity; // Sets the verbosity level.
use easy_ga::LOG_verbosity_type; // Sets the verbosity type.

LOG_verbosity(VerbosityLevel::LOW); // VerbosityLevel::DISABLED by default
LOG_verbosity_type(VerbosityType::LOG_AND_SAVE); // VerbosityType::LOG by default
```

- **VerbosityLevel:**
  - **DISABLED**: The logs are disabled.
  - **LOW**: Only very usefull information.
  - **MID**: Maybe not to desired information but also usefull.
  - **HIGH**: All logs are avaliable including tracing logs.

- **VerbosityType:**
  - **LOG**: Only terminal logs.
  - **SAVE**: Saves the logs into [target/easy_ga/logs/](target/easy_ga/logs/).
  - **SAVE_AND_LOG**: Both.

---

## Benchmarking

Benchmarking was added in the version `1.1.0` and you can run them donwloading the repository and running `cargo bench` from the command-line. The benchmarks are placed inside the [benches/](benches/) folder. 

---

## Next steps

This is a personal side project mainly for me so any further implementations will be done in my spare time as a good way to teach me more about Rust.

- Multithreading
- Add verbosity for debugging ✅
- More unit testing and system testing
- New default `Selection` algorithms
- CSV and JSON result export
- Fix some quality of life problems with references and chain calling
- Add benchmarks ✅

## License

Easy_GA is licensed under Mozilla Public License 2.0.


CHANGELOG.md
# [1.2.0]

* Added new `SelectionAlgorithm::Random`.
* Added new `SelectionAlgorithm::Stochastic`.
* `SelectionAlgorithm` now derive `PartialEq`, `Eq`, `PartialOrd` and `Ord`.
* Added unit testing for `Selection` trait.
* Logs are now saved inside `target/easy_ga/logs` insted of `target/tmp/`

# [1.1.0]

* Added benchmarks. You can run benchmarks using `cargo bench`. The benches are all inside `benches/`.
* Added `logger` feature to get information about the execution.
* Added `logger::VerbosityLevel` to filter the output.
* Added `logger::VerbosityType` to choose between only LOG, SAVE and LOG_AND_SAVE. 

benches/bench_genetic_algorithm.rs
pub mod benchmark {
    use criterion::BenchmarkId;
    use criterion::{black_box, Criterion};
    use easy_ga::samples::MyGene as MockMyGene;
    use easy_ga::GeneticAlgorithm;
    use easy_ga::LOG_verbosity;
    use easy_ga::LOG_verbosity_type;
    use easy_ga::SelectionAlgorithms;
    use easy_ga::VerbosityLevel;
    use easy_ga::VerbosityType;

    /// Benchmark the GeneticAlgorithm::new function.
    pub fn new(c: &mut Criterion) {
        c.bench_function("GeneticAlgorithm::new", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new();
                }
            })
        });
    }

    /// Benchmark the GeneticAlgorithm::new_with_values function.
    pub fn new_with_values(c: &mut Criterion) {
        let mut group = c.benchmark_group("GeneticAlgorithm::new_with_values");

        group.bench_function("Tournament", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new_with_values(
                        black_box(100),
                        black_box(1000),
                        black_box(0.05),
                        black_box(0.90),
                        Box::new(SelectionAlgorithms::Tournament(black_box(2))),
                        black_box(f64::MAX),
                    );
                }
            })
        });

        group.bench_function("Roulette", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new_with_values(
                        black_box(100),
                        black_box(1000),
                        black_box(0.05),
                        black_box(0.90),
                        Box::new(SelectionAlgorithms::Roulette),
                        black_box(f64::MAX),
                    );
                }
            })
        });

        group.finish();
    }

    /// Benchmarks the initialization of a GeneticAlgorithm with a chain-calling method.
    pub fn new_chain_calling(c: &mut Criterion) {
        c.bench_function("GeneticAlgorithm::new_chain_calling", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new()
                        .population_size(black_box(100))
                        .iterations(black_box(1000))
                        .mutation_rate(black_box(0.05))
                        .selection_rate(black_box(0.90))
                        .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(
                            2,
                        ))))
                        .fitness_goal(black_box(f64::MAX));
                }
            })
        });
    }

    /// Benchmarks a comparation between the 3 ways for instantiate a GeneticAlgorithm.
    pub fn genetic_algorithm_instantiation_comparation(c: &mut Criterion) {
        let mut group = c.benchmark_group("GeneticAlgorithm instantiation");

        group.bench_function("new()", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new();
                }
            })
        });

        group.bench_function("new_with_values()", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new_with_values(
                        black_box(100),
                        black_box(1000),
                        black_box(0.05),
                        black_box(0.90),
                        Box::new(SelectionAlgorithms::Tournament(black_box(2))),
                        black_box(f64::MAX),
                    );
                }
            })
        });

        group.bench_function("chain_calling", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    GeneticAlgorithm::<MockMyGene>::new()
                        .population_size(black_box(100))
                        .iterations(black_box(1000))
                        .mutation_rate(black_box(0.05))
                        .selection_rate(black_box(0.90))
                        .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(
                            2,
                        ))))
                        .fitness_goal(black_box(f64::MAX));
                }
            })
        });

        group.finish();
    }

    /// Benchmark the GeneticAlgorithm::init function.
    pub fn init(c: &mut Criterion) {
        c.bench_function("GeneticAlgorithm::init", |b| {
            b.iter(|| {
                for _ in 0..10000 {
                    let _ = GeneticAlgorithm::<MockMyGene>::new().init();
                }
            })
        });
    }

    /// Benchmarks the GeneticAlgorithm::run method with different population sizes.
    pub fn run_different_population_size(c: &mut Criterion) {
        let mut group = c.benchmark_group("GeneticAlgorithm::run - population_size");
        group.measurement_time(core::time::Duration::from_secs(10));
        let population_sizes = [10, 50, 100, 250, 500];

        for population_size in population_sizes {
            group.bench_with_input(
                BenchmarkId::from_parameter(population_size),
                &population_size,
                |b, &population_size| {
                    b.iter(|| {
                        GeneticAlgorithm::<MockMyGene>::new()
                            .population_size(black_box(population_size))
                            .iterations(black_box(1000))
                            .mutation_rate(black_box(0.05))
                            .selection_rate(black_box(0.90))
                            .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(
                                black_box(2),
                            )))
                            .fitness_goal(black_box(f64::MAX))
                            .init()
                            .run();
                    })
                },
            );
        }

        group.finish();
    }

    /// Benchmarks the logger with VerbosityType::LOG.
    pub fn logger_log_with_different_verbosity(c: &mut Criterion) {
        let mut group = c.benchmark_group("Logger_LOG - VerbosityLevel::");
        group.measurement_time(core::time::Duration::from_secs(10));

        LOG_verbosity_type(VerbosityType::LOG);

        LOG_verbosity(VerbosityLevel::DISABLED);
        group.bench_function("DISABLED", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::LOW);
        group.bench_function("LOW", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::MID);
        group.bench_function("MID", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::HIGH);
        group.bench_function("HIGH", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        group.finish();
    }

    /// Benchmarks the logger with VerbosityType::SAVE.
    pub fn logger_save_with_different_verbosity(c: &mut Criterion) {
        let mut group = c.benchmark_group("Logger_SAVE - VerbosityLevel::");
        group.measurement_time(core::time::Duration::from_secs(10));

        LOG_verbosity_type(VerbosityType::SAVE);

        LOG_verbosity(VerbosityLevel::DISABLED);
        group.bench_function("DISABLED", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::LOW);
        group.bench_function("LOW", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::MID);
        group.bench_function("MID", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::HIGH);
        group.bench_function("HIGH", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        group.finish();
    }

    /// Benchmarks the logger with VerbosityType::SAVE.
    pub fn logger_log_and_save_with_different_verbosity(c: &mut Criterion) {
        let mut group = c.benchmark_group("Logger_LOG_AND_SAVE - VerbosityLevel::");
        group.measurement_time(core::time::Duration::from_secs(10));

        LOG_verbosity_type(VerbosityType::LOG_AND_SAVE);

        LOG_verbosity(VerbosityLevel::DISABLED);
        group.bench_function("DISABLED", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::LOW);
        group.bench_function("LOW", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::MID);
        group.bench_function("MID", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        LOG_verbosity(VerbosityLevel::HIGH);
        group.bench_function("HIGH", |b| {
            b.iter(|| {
                GeneticAlgorithm::<MockMyGene>::new()
                    .population_size(black_box(20))
                    .iterations(black_box(100))
                    .mutation_rate(black_box(0.05))
                    .selection_rate(black_box(0.90))
                    .selection_algorithm(Box::new(SelectionAlgorithms::Tournament(black_box(2))))
                    .fitness_goal(black_box(f64::MAX))
                    .init()
                    .run();
            })
        });

        group.finish();
    }
}
